torch/csrc/utils/tensor_flatten.h

```diff
@@ -37,14 +41,35 @@ struct TensorGroup {
   }
 };
 
-std::vector<TensorGroup> take_tensors(at::TensorList tensors, size_t size_limit);
-void reorder_tensors_like(std::vector<at::Tensor>& tensors, at::TensorList order);
+// Helper function that take a list of tensor by the size limit, if the
+// list of tensors are of different tensor types, each type will be taken by
+// the size limit. In other words, the input tensor list will firstlbe bucketed
+// by tensor types, and then each bucket will be further bucketed by size_limit
+std::vector<TensorGroup> take_tensors(
+    at::TensorList tensors,
+    size_t size_limit);
+
+// Helper function that take a list of tensor by the size limit in a
+// fine-grained manner. This helper will loop through the input tensor list and
+// accumulate enough tensors (even though the tensors can be of different type)
+// until size_limit is hit. Then for each accumulation, it will further bucket
+// the accumulation by different types. In other words, for each size_limit's
+// accumulation, multiple buckets are generated.
```

Comment:
Multiple buckets as in 1 per type (in the last sentence)? If I understand correctly this will output in chronological order (since you don't care about type when counting up to the size limit), compared to potential complete out of order for the other implementation. If this is the case, I would assume this more fine grained one has a much better shot overlapping indeed and we should consider just replacing it instead of having 2 functions with similar outcome.

Reply:
Well, hold on, the DDP backward bucketing is using the Python implementation of take_tensors, which is different than C++ :D.

Let's keep two for now, I think both can be useful in certain cases.

Reply:
So this basically swaps the order of bucketing compared to the function above? Why would anyone want that?

Reply:
This would be beneficial for BW overlapping, especially for the case of fp16/fp32 mixed precision training.